src/hotspot/share/cds/archiveUtils.hpp

```diff
@@ -291,4 +293,79 @@ class HeapRootSegments {
   HeapRootSegments& operator=(const HeapRootSegments&) = default;
 };
 
+class ArchiveWorkers;
+
+// A task to be worked on by worker threads
+class ArchiveWorkerTask : public CHeapObj<mtInternal> {
+  friend class ArchiveWorkers;
+private:
+  const char* _name;
+  int _max_chunks;
+  volatile int _chunk;
+
+  void run();
+
+  void maybe_override_max_chunks(int max_chunks);
+
+public:
+  ArchiveWorkerTask(const char* name, int max_chunks = -1) :
+      _name(name), _max_chunks(max_chunks), _chunk(0) {}
+  const char* name() const { return _name; }
+  virtual void work(int chunk, int max_chunks) = 0;
+};
+
+class ArchiveWorkerThread : public NamedThread {
+  friend class ArchiveWorkers;
+private:
+  ArchiveWorkers* const _pool;
+
+public:
+  ArchiveWorkerThread(ArchiveWorkers* pool);
+  const char* type_name() const override { return "Archive Worker Thread"; }
+  void run() override;
+};
+
+class ArchiveWorkerShutdownTask : public ArchiveWorkerTask {
+public:
+  ArchiveWorkerShutdownTask() : ArchiveWorkerTask("Archive Worker Shutdown", 1) {}
+  void work(int chunk, int max_chunks) override {
+    // Do nothing.
+  }
+};
+
+// Special worker pool for archive workers. The goal for this pool is to
+// startup fast, distribute spiky workloads efficiently, and being able to
+// shutdown after use. This makes the implementation quite different from
+// the normal GC worker pool.
+class ArchiveWorkers {
+  friend class ArchiveWorkerThread;
+private:
+  // The reciprocal ratio for number of workers per CPU. We are targeting
+  // to take 1/4 CPUs to provide decent parallelism without letting workers
+  // stumble over each other.
+  static constexpr int cpus_per_worker = 4;
```

Comment:
Could these static constexpr values maybe be declared either with am ALL_CAPS_NAME or _leading_underscore. They currently look like locals/params at point of use.

Reply:
Done in new commit. (I'll push all them soon.)